### ISSUE TITLE
Debugging non-secure code under GNU/Eclipse does not work

### ISSUE THREAD

**Issue by 0xd35a2ee** [OpenNuvoton/M2351BSP#2]
Hello there,

debugging non-secure code with GNU/Eclipse nad OpenOCD under Linux does not seem to work. Building the code is not an issue, but flashing it on the board fails (simply hangs) with the following error message:
```
Info : Nuvoton NuMicro: Sector Erase ... (131200 to 131205)
Info : Nuvoton NuMicro: Flash Write ...
Warn : keep_alive() was not invoked in the 1000ms timelimit. GDB alive packet not sent! (1574). Workaround: increase "set remotetimeout" in GDB
```
On the other hand, building and debugging secure code works like a charm.
```
Info : 0x00300000: 0x00000000
Info : 0x00300000: 0xffffffff
Info : 0x00300004: 0xffffffff
Info : 0x00300008: 0xffffffff
Info : 0x0030000c: 0xffffffff
Info : 0x00200800: 0x00000000
Info : 0x00210800: 0x00000000
Info : 0x00200800: 0x00040000
Info : 0x00200804: 0x0000005a
Info : 0x00210804: 0x0000005a
auto erase enabled
Info : Nuvoton NuMicro: Sector Erase ... (0 to 6)
Info : Nuvoton NuMicro: Flash Write ...
Info : Have written 28%
Info : Have written 57%
Info : Have written 85%
Info : Have written 100%
Info : Nuvoton NuMicro: Sector Erase ... (126 to 126)
Info : Nuvoton NuMicro: Flash Write ...
Info : Have written 100%
wrote 16384 bytes Secure/Debug/Secure.hex in 4.195044s (3.814 KiB/s)
NuMicro.cpu: target state: halted
target halted due to debug-request, current mode: Thread
```

The code that was used for this demonstration is the TrustZone template code, which is generated by GNU/Eclipse automatically after creating a secure and non-secure project respectively. The configurations steps were followed as described in the official GNU/Eclipse user manual. The only difference is that from the secure zone I explicitly set `NCSBA = 0x040000` and `SCRLOCK = 0x5A` and `ARLOCK = 0x5A` because by default `NCSBA = 0x080000` would be written, which does not correspond to the default partitioning `#define FMC_SECURE_ROM_SIZE      0x40000.` Such a mismatch  causes the device to [hang in an infinite loop](https://github.com/OpenNuvoton/M2351BSP/blob/master/SampleCode/TrustZone/Template/partition_M2351.h#L99).

**Comment by 0xd35a2ee** [comment]
After some investigation I found that if I lock the *secure* flash memory region, using for example `SCRLOCK = 0x11`, I am able to debug *non-secure* code as expected. However, now debugging *secure* code is not possible because of the lock. The chip must be totally erased (e.g. using the windows ICP tool) to enable *secure* debugging again. Erasing the chip through GNU/Eclipse by checking the `Erase chip` box in the debug configurations does not work. 

How can simultaneous debugging of secure and non-secure code be achieved, without the need of additional locks/erasures?

**Comment by wschang0** [comment]
1.     Open M2351_NS Project

2.     checking the erase chip box and Non Secure box in the debug configurations

3.     Apply settings and enter debug mode, the chip will be totally erased



===========================================================================================
The privileged confidential information contained in this email is intended for use only by the addressees as indicated by the original sender of this email. If you are not the addressee indicated in this email or are not responsible for delivery of the email to such a person, please kindly reply to the sender indicating this fact and delete all copies of it from your computer and network server immediately. Your cooperation is highly appreciated. It is advised that any unauthorized use of confidential information of Nuvoton is strictly prohibited; and any information in this email irrelevant to the official business of Nuvoton shall be deemed as neither given nor endorsed by Nuvoton.

**Comment by wschang0** [comment]
The wrong NSCBA setting will cause the infinite loop.
It is necessary to avoid wrong TrustZone partition.
you should full erase the whole chip to fix it.

**Comment by 0xd35a2ee** [comment]
Thanks for your answer @wschang0. Follow up after trying what you have described.

1. How should the debugging normally work? Is the concept that the user can (1) **simultaneously** debug secure and non-secure code by jumping from one zone to the other or (2) is it intended that the user debugs secure and non-secure code **separately** from one another?
You suggested erasing the chip while debugging non-secure code. This leads to me to think that a separate debugging is intended. 

2. I tried what you have suggested (marking the chip erase and non-secure boxes in the debug configuration). This sadly does not resolve the problem. 

I left the flasher to run for more than 2 minutes and it actually managed to flash the non-secure code in flash memory with very low speed.
```
Warn : keep_alive() was not invoked in the 1000ms timelimit. GDB alive packet not sent! (1576). Workaround: increase "set remotetimeout" in GDB
Error: timed out while waiting for target halted
Error: Error executing NuMicro Flash programming algorithm
wrote 11260 bytes from file /home/user/project/NonsecureWorld/Debug/NonsecureWorld.hex in 101.134171s (0.109 KiB/s)
```
However, the chip did not seem to be executing the correct code (no LEDs were blinking). I paused the debugger and saw that it executes code from the address `0xeffffffe`, which is invalid. Hence I suspect that there is a problem with the memory partitioning and I cannot find it. I attach the linker scripts, compiled hex files and the memory partition header file.

The secure code debug configuration:
![secure-world-debug-config](https://user-images.githubusercontent.com/43346167/47003219-0efe1200-d12f-11e8-9cb8-e50a07850fb1.png)

The non-secure debug configuration:
![nonsecure-world-debug-config](https://user-images.githubusercontent.com/43346167/47003218-0efe1200-d12f-11e8-9e8c-ed122811d5f4.png)

[secure-linker-script.ld.txt](https://github.com/OpenNuvoton/M2351BSP/files/2482131/secure-linker-script.ld.txt)

[non-secure-linker-script.ld.txt](https://github.com/OpenNuvoton/M2351BSP/files/2482136/non-secure-linker-script.ld.txt)

[partition_M2351.h.txt](https://github.com/OpenNuvoton/M2351BSP/files/2482144/partition_M2351.h.txt)

[SecureWorld.hex.txt](https://github.com/OpenNuvoton/M2351BSP/files/2482149/SecureWorld.hex.txt)

[NonsecureWorld.hex.txt](https://github.com/OpenNuvoton/M2351BSP/files/2482151/NonsecureWorld.hex.txt)


Where is the problem?

**Comment by wschang0** [comment]
You may try Keil Nuvoton edition first to see what is normal in working with TrustZone.
The tool is free and could be download from:
https://m2351.nuvoton.com/resource/

**Comment by 0xd35a2ee** [comment]
Since the main problem was neglected and I found a solution on my own, I will post it here with the hope that it will be helpful for others:

**TL;DR. Deselect the non-secure option in the debug configurations of the non-secure codebase.**

A more detailed explanation:
**Step 0.** Reinstall the Nuvoton version of GNU/Eclipse. In my case this was needed because the global settings were polluted after numerous different attempts to resolve the issue.

**Step 1.** Create a new **secure** project. Make sure you give the correct flash/SRAM size, which will be used in `partition_M2351.h`. Be default this is `32KB` SRAM and `256KB` flash.

**Step 2.** Build the **secure** project and create a new debug profile. Besides everything which is explained in the official GNU/Eclipse manual from Nuvoton, the following settings will work with the default configurations:

![secure](https://user-images.githubusercontent.com/43346167/47078983-be58e880-d204-11e8-8092-d7690edde8b7.png)

**Step 3.** Create a new **non-secure** project. Make sure you give the correct flash/SRAM size, which you have already specified in `partition_M2351.h` that is part of the **secure** codebase. By default the non-secure world will be using `64KB` SRAM and `256KB` flash. Leave the library path empty.

**Step 4.** For the **non-secure** project add the path to the **secure** library (this is normally *SecureProjectName/Debug/Library*) from the *Project Settings > C/C++ > Build > GNU Linker Libraries*. After that build the **non-secure** project.

**Step 5.** Create a new debug configuration for the **non-secure** project. The configurations should be as on the picture below. Do not check the non-secure checkbox because this creates problems as described in the above comments.
 
![non-secure](https://user-images.githubusercontent.com/43346167/47078945-ad0fdc00-d204-11e8-81f8-d3107c2daa85.png)
